### ISSUE TITLE
Elem-wise Add/Time Forward

### ISSUE THREAD

**Issue by silentspring2** [acada-sjtu/EdgeTraining#41]
__TODO:__
- [ ] Elem-wise Add
    - [ ] NNVM level
    - [ ] TOPI LLVM
    - [ ] TOPI Mali
- [ ] Elem-wise Time
    - [ ] NNVM level
    - [ ] TOPI LLVM
    - [ ] TOPI Mali

--------
__Principles__
1. We will define a softbound which drops certain information while merging as previously discussed.
2. If time is limited, we would mainly focus on Add. The Time Op is very much like the Add Op.

**Comment by silentspring2** [comment]
Need algorithm level support to define how to drop information while merging so that accuracy loss can be minimized.

**Comment by ACA-Lab-SJTU** [comment]
Let’s discuss together this afternoon. 

-
Li Jiang, Ph.D., Associate Professor,
Department of computer science and engineering,
Shanghai Jiao Tong University

**Comment by rogercmq** [comment]
For networks with residual blocks, we may only prune the first two layers of each block in ResNet for simplicity, leaving the block output and projection shortcuts unchanged, which is implemented in [1]. I will do some survey on network compression methods for element-wise OPs.

Reference:
[1] ThiNet: A Filter Level Pruning Method for Deep Neural Network Compression, ICCV 2017

**Comment by silentspring2** [comment]
I quickly go through these two papers. It is a smart way to do channel pruning like this to avoid the system level re-implementation. In fact, they are building smaller networks.

But given we are a system-level framework, I suggest that we might need more general support than these methods. Unless it can be demonstrated that these methods are widely used and treated as common sense in industry and academic fields.

**Comment by ACA-Lab-SJTU** [comment]
Agreed. But, put it in mind that we cannot build Rome in one day. If this issue is not significant, we should propose our paper ASAP and leave the re-implementation later. 

-
Li Jiang, Ph.D., Associate Professor,
Department of computer science and engineering,
Shanghai Jiao Tong University

**Comment by silentspring2** [comment]
If we wish to publish the first version quickly, I suggest we follow @rogercmq 's idea, which is assuming __no channel pruning is implemented on shortcut__. The following are the reasons:
1. The FLOPs reduction is not significant. If we remove half of the channels in a resnet block just like Fig.3 of ThiNet. The FLOPs reduction is 61.2% and 75% respectively before and after removing this assumption. 
2. The implementation might be hard. If we adopt the assumption, we can copy the shortcut to the output tensor and add the corresponding data to the output tensor. The output index tensor can be replaced with a serial easily. However, if we do not adopt this assumption, we will create an empty output and output index tensor. For each channel, we will encounter four situations:
    1. None input channel matches
    2. Channel 1 matches
    3. Channel 2 matches
    4. Both channel 1 and 2 match
The control flow might be more complicated.
